eth/vm/base.py

```diff
@@ -151,6 +155,10 @@ def build_state(cls,
         execution_context = cls.create_execution_context(header, previous_hashes, chain_context)
         return cls.get_state_class()(db, execution_context, header.state_root)
 
+    @cached_property
+    def _consensus(self) -> ConsensusAPI:
+        return self.consensus_class(self.chaindb.db)
```

Comment:
Doesn't this mean each VM would get it's own instance of `consensus_class`? That would be tragic as the `ConsensusAPI` can be stateful (Clique keeps an in-memory cache). We currently do not have a test case for this but it would be good to add one. It should be easy to modify [this](https://github.com/ethereum/py-evm/pull/1875/files#diff-dbda3a284b12bde27a15171ceed1dbc3R28) to keep state in memory and mine blocks that pass over VM boundaries.

Reply:
Yes, and I think it is important to figure out how to make it stateless. Everything else that might change in a fork (in the VM) can be built up and torn down on demand, which we lean on heavily to ensure smooth fork transitions. Managing a change in Clique at a block boundary will be a nightmare if it requires some persistent in-memory store.

There must be some way for Clique to load the data from disk. How long does that take in practice?

Reply:
Mmmh..ok let's see. So, there are multiple places of caching in the current clique implementation but let's ignore the one's that are for performance and look at the [`HeaderCache`](https://github.com/ethereum/py-evm/blob/master/eth/consensus/clique/header_cache.py) which I think is the problematic one.

To recall, in Clique you can not validate the seal of a header unless you have access to the previous header. Let's assume we validate a series of headers via `validate_chain(...)`. These headers may span over multiple VMs and result in a `validate_seal(...)` call on each of these VMs. By the time that we validate the seal of the second header (or third, fourth, `n`s) the previous header won't be in the database yet (after all we are just in the middle of validating a whole series so non of it will get to the db until we've validated the entire series). So in order to look it up, we turn to our `HeaderCache` which we flush out as soon as we know that the headers are in the db (we actually keep the last `1000` in memory to account for temporary forks).

https://github.com/ethereum/py-evm/blob/843ab2969e2c47678185c940ea3a58d7693d453d/eth/consensus/clique/snapshot_manager.py#L72-L80

The other way to fix that that I see is to make `validate_chain` clique aware because that method has access to the entire series but then this would make the consensus engine operate on the chain level again.